### PR TITLE
Migrate shared memory code

### PR DESCRIPTION
**Pull request by PointKernel**
Depends on #403

This PR migrates the shared memory tests to exercise the new experimental map.

### CODE REVIEW

**Comment by PointKernel** [comment]
Ideally, `make_copy` should be available only if query operators are present and `initialize` can be exclusively invoked from insert refs. However, there is no such "category" concept in the current design. This PR is mainly to make the code work and the followup API trimming or cleanups are necessary.

Those two functions are no longer "static" simply because I haven't figured out how to pass the variadic operator parameters via a static function. Is that even possible?

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 1]
```diff
@@ -268,6 +278,63 @@ class open_addressing_ref_impl {
    */
   [[nodiscard]] __host__ __device__ constexpr iterator end() noexcept { return storage_ref_.end(); }
 
+  /**
+   * @brief Makes a copy of the current device ref using non-owned memory
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
   * @brief Makes a copy of the current device reference using non-owned memory.
```

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 2]
```diff
@@ -268,6 +278,63 @@ class open_addressing_ref_impl {
...
+    window_type const* const windows_ptr = this->storage_ref().data();
+    for (size_type i = g.thread_rank(); i < num_windows; i += g.size()) {
+      memory_to_use[i] = windows_ptr[i];
+    }
+    g.sync();
+#endif
+  }
+
+  /**
+   * @brief Initializes the container storage using the threads in the group `tile`.
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
   * @brief Initializes the container storage.
```

File: include/cuco/static_map_ref.cuh [hunk 3]
```diff
@@ -166,6 +166,20 @@ class static_map_ref
...
+   * @return The sentinel value used to represent an erased key slot
+   */
+  [[nodiscard]] __host__ __device__ constexpr key_type erased_key_sentinel() const noexcept;
+
+  /**
+   * @brief Gets the key comparator.
+   *
+   * @return The comparator used to compare keys
+   */
+  [[nodiscard]] __host__ __device__ constexpr key_equal key_eq() const noexcept;
```
Comment (sleeepyjack) [inline include/cuco/static_map_ref.cuh:181]: STL also exposes both `key_eq` and the hash function. We should do the same for the new set/map and the corresponding refs for API parity reasons. Could be another PR.

File: include/cuco/static_set_ref.cuh [hunk 4]
```diff
@@ -161,6 +175,36 @@ class static_set_ref
   template <typename... NewOperators>
   [[nodiscard]] __host__ __device__ auto with(NewOperators... ops) && noexcept;
 
+  /**
+   * @brief Makes a copy of the current device ref using non-owned memory
```
Comment (sleeepyjack) [inline include/cuco/static_set_ref.cuh]:
```suggestion
   * @brief Makes a copy of the current device reference using non-owned memory.
```

File: include/cuco/static_map_ref.cuh [hunk 5]
```diff
@@ -183,6 +197,36 @@ class static_map_ref
   template <typename... NewOperators>
   [[nodiscard]] __host__ __device__ auto with(NewOperators... ops) && noexcept;
 
+  /**
+   * @brief Makes a copy of the current device ref using non-owned memory
+   *
+   * This function is intended to be used to create shared memory copies of small static maps,
+   * although global memory can be used as well.
+   *
+   * @tparam CG The type of the cooperative thread group
```
Comment (sleeepyjack) [inline include/cuco/static_map_ref.cuh]:
```suggestion
   * although global memory can be used as well.
   *
   * @note This function synchronizes the group `tile`.
   *
   * @tparam CG The type of the cooperative thread group
```

File: include/cuco/static_set_ref.cuh [hunk 6]
```diff
@@ -161,6 +175,36 @@ class static_set_ref
...
+   *
+   * This function is intended to be used to create shared memory copies of small static sets,
+   * although global memory can be used as well.
+   *
+   * @tparam CG The type of the cooperative thread group
```
Comment (sleeepyjack) [inline include/cuco/static_set_ref.cuh]:
```suggestion
   * although global memory can be used as well.
   *
   * @note This function synchronizes the group `tile`.
   *
   * @tparam CG The type of the cooperative thread group
```

File: tests/static_map/shared_memory_test.cu [hunk 7]
```diff
@@ -93,31 +98,35 @@ TEMPLATE_TEST_CASE_SIG("Shared memory static map",
...
-    maps.push_back(std::make_unique<MapType>(
-      map_capacity, cuco::empty_key<Key>{-1}, cuco::empty_value<Value>{-1}));
+    maps.push_back(std::make_unique<map_type>(
+      extent_type{}, cuco::empty_key<Key>{-1}, cuco::empty_value<Value>{-1}));
   }
 
   thrust::device_vector<bool> d_keys_exist(number_of_maps * elements_in_map);
   thrust::device_vector<bool> d_keys_and_values_correct(number_of_maps * elements_in_map);
 
+  using ref_type = decltype(maps.front()->ref(cuco::experimental::op::insert));
```
Comment (sleeepyjack) [inline tests/static_map/shared_memory_test.cu]: Can we instead use `map_type::ref_type`, i.e., ref without any operators and then in the kernel dispatch the ref with `insert`?

File: include/cuco/static_map_ref.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/static_map_ref.cuh:181]: Right. The only reason that I choose not to expose the hash function is "double hashing". :(

File: tests/static_map/shared_memory_test.cu (hunk 7 above)
Comment (PointKernel) [reply tests/static_map/shared_memory_test.cu]:
```cpp
using ref_type = typename map_type::ref_type<cuco::experimental::op::insert_tag>;
```
Cannot be better. Good call!

**Review by sleeepyjack — APPROVED** [review]
LGTM 🚀
